### PR TITLE
LOG-3282: Generate reference documentation for apis

### PR DESCRIPTION
**Pull request by jcantrill**
### Description
This PR:

* Adds reference api documentation that is generated based upon go structs

### Links
* https://issues.redhat.com/browse/LOG-3282

### CODE REVIEW

**Review by alanconway — CHANGES_REQUESTED** [review]
This looks great, esp. synchronizing the data model with the code. The patch below has minor cosmetic tweaks to the adoc templates. We need to review the data model section but I think you should merge this as-is and refine as we go - this is already better than what we had before.

```
From 34541cb7ac26b167588d78cf870e756501c83c3c Mon Sep 17 00:00:00 2001
From: Alan Conway <aconway@redhat.com>
Date: Mon, 9 Jan 2023 14:40:54 -0500
Subject: [PATCH] Diffs for Jeffs data model work.

---
 Makefile                                      |  4 ++--
 config/docs/templates/apis/asciidoc/pkg.tpl   | 19 ++++++++--------
 .../templates/datamodels/asciidoc/pkg.tpl     | 22 ++++++++++---------
 docs/reference/datamodels/viaq/v1.adoc        |  9 ++++----
 docs/reference/operator/api.adoc              | 12 +++++-----
 5 files changed, 34 insertions(+), 32 deletions(-)

diff --git a/Makefile b/Makefile
index 18517ce17..9f2a957d7 100644
--- a/Makefile
+++ b/Makefile
@@ -121,11 +121,11 @@ build-debug:
 docs: docs/reference/operator/api.adoc docs/reference/datamodels/viaq/v1.adoc
 .PHONY: docs
 
-docs/reference/operator/api.adoc:
+docs/reference/operator/api.adoc: $(GEN_CRD_API_REFERENCE_DOCS)
 	$(GEN_CRD_API_REFERENCE_DOCS) -api-dir "github.com/openshift/cluster-logging-operator/apis/" -config "$(PWD)/config/docs/config.json" -template-dir "$(PWD)/config/docs/templates/apis/asciidoc" -out-file "$(PWD)/$@"
 .PHONY: docs/reference/operator/api.adoc
 
-docs/reference/datamodels/viaq/v1.adoc:
+docs/reference/datamodels/viaq/v1.adoc: $(GEN_CRD_API_REFERENCE_DOCS)
 	$(GEN_CRD_API_REFERENCE_DOCS) -api-dir "github.com/openshift/cluster-logging-operator/internal/datamodels/viaq/v1" -config "$(PWD)/config/docs/config.json" -template-dir "$(PWD)/config/docs/templates/datamodels/asciidoc" -out-file "$(PWD)/$@"
 .PHONY: docs/reference/datamodels/viaq/v1.adoc
 
diff --git a/config/docs/templates/apis/asciidoc/pkg.tpl b/config/docs/templates/apis/asciidoc/pkg.tpl
index 8a6915e8e..882a7b56e 100644
--- a/config/docs/templates/apis/asciidoc/pkg.tpl
+++ b/config/docs/templates/apis/asciidoc/pkg.tpl
@@ -5,15 +5,16 @@
 toc::[]
 
 {{ range .packages -}}
+= Package {{ packageDisplayName . }}
 
-        {{- range (sortedTypes (visibleTypes .Types )) -}}
-            {{- if isObjectRoot . -}}
+    {{range (sortedTypes (visibleTypes .Types )) -}}
+        {{if isObjectRoot . -}}
 
-            == {{ (typeDisplayName .) }}
-            {{  (comments .CommentLines) }}
-            {{- template "type" (nodeParent . "") -}}
+== {{ (typeDisplayName .) }}
+{{ (comments .CommentLines) }}
+{{ template "type" (nodeParent . "") -}}
 
-            {{- end -}}
-        {{- end -}}
-{{- end -}}
-{{- end -}}
\ No newline at end of file
+        {{end -}}
+    {{end -}}
+{{end -}}
+{{end -}}
diff --git a/config/docs/templates/datamodels/asciidoc/pkg.tpl b/config/docs/templates/datamodels/asciidoc/pkg.tpl
index 8a6915e8e..95f2947e7 100644
--- a/config/docs/templates/datamodels/asciidoc/pkg.tpl
+++ b/config/docs/templates/datamodels/asciidoc/pkg.tpl
@@ -1,19 +1,21 @@
 {{- define "packages" -}}
+= Logging Data Model Reference
 :toc:
 :toclevels: 2
-:toc-placement!:
-toc::[]
+:doctype: book
 
 {{ range .packages -}}
+= Package {{ packageDisplayName . }}
 
-        {{- range (sortedTypes (visibleTypes .Types )) -}}
-            {{- if isObjectRoot . -}}
+    {{ range (sortedTypes (visibleTypes .Types )) -}}
+	{{- if isObjectRoot . -}}
 
-            == {{ (typeDisplayName .) }}
-            {{  (comments .CommentLines) }}
-            {{- template "type" (nodeParent . "") -}}
+== {{ (typeDisplayName .) }}
+{{  (comments .CommentLines) }}
+{{- template "type" (nodeParent . "") -}}
 
-            {{- end -}}
-        {{- end -}}
+	{{- end -}}
+    {{- end -}}
+
+{{- end -}}
 {{- end -}}
-{{- end -}}
\ No newline at end of file
diff --git a/docs/reference/datamodels/viaq/v1.adoc b/docs/reference/datamodels/viaq/v1.adoc
index 3ad7160d2..be3f07720 100644
--- a/docs/reference/datamodels/viaq/v1.adoc
+++ b/docs/reference/datamodels/viaq/v1.adoc
@@ -1,10 +1,11 @@
+= Logging Data Model Reference
 :toc:
 :toclevels: 2
-:toc-placement!:
-toc::[]
+:doctype: book
 
-== Viaq Data Model for kubernetes api events
+= Package viaq/v1
 
+== Viaq Data Model for kubernetes api events
 The data model for collected audit event logs from kubernetes or OpenShift api servers.
 
 nolint:govet
@@ -947,7 +948,6 @@ in Elasticsearch, this field will give you the exact document corresponding to t
 * string
 
 == Viaq Data Model for Containers
-
 The data model for collected logs from containers.
 
 [options="header"]
@@ -1691,7 +1691,6 @@ The `structured` field includes the same sub-fields as the original log message.
 * object
 
 == Viaq Data Model for journald
-
 The data model for collected logs from node journal.
 
 [options="header"]
diff --git a/docs/reference/operator/api.adoc b/docs/reference/operator/api.adoc
index d3e8afccd..4c97f141d 100644
--- a/docs/reference/operator/api.adoc
+++ b/docs/reference/operator/api.adoc
@@ -3,8 +3,9 @@
 :toc-placement!:
 toc::[]
 
-== ClusterLogForwarder
+= Package logging.openshift.io/v1
 
+== ClusterLogForwarder
 ClusterLogForwarder is an API to configure forwarding logs.
 
 You configure forwarding by specifying a list of `pipelines`,
@@ -18,7 +19,7 @@ you can define your own outputs with a URL and other connection information
 to forward logs to other stores or processors, inside or outside the cluster.
 
 For more details see the documentation on the API fields.
-
+
 [options="header"]
 |======================
 |Property|Type|Description
@@ -115,7 +116,8 @@ A label selector is a label query over a set of resources.
 
 === .spec.outputDefaults.elasticsearch
 ===== Description
-
+ElasticsearchStructuredSpec is spec related to structured log changes to determine the elasticsearch index
+
 =====  Type
 * object
 
@@ -126,7 +128,6 @@ A label selector is a label query over a set of resources.
 |enableStructuredContainerLogs|bool|  *(optional)* EnableStructuredContainerLogs enables multi-container structured logs to allow
 |structuredTypeKey|string|  *(optional)* StructuredTypeKey specifies the metadata key to be used as name of elasticsearch index
 |structuredTypeName|string|  *(optional)* StructuredTypeName specifies the name of elasticsearch schema
-|version|int|  *(optional)* Version specifies the version of Elasticsearch to be used.
 |======================
 
 === .spec.outputs[]
@@ -221,9 +222,8 @@ PipelinesSpec link a set of inputs to a set of outputs.
 * array
 
 == ClusterLogging
-
 A Red Hat OpenShift Logging instance. ClusterLogging is the Schema for the clusterloggings API
-
+
 [options="header"]
 |======================
 |Property|Type|Description
-- 
2.39.0

```

**Comment by jcantrill** [comment]
@alanconway is this your intent of how the datamodel displays, without being able to link directly to a path?

image

File: apis/logging/v1/cluster_log_forwarder_types.go
Comment (jcantrill) [reply apis/logging/v1/cluster_log_forwarder_types.go:148]: We could but not harming anything ATM.  I leave it to you if I remove them in this PR

**Comment by jcantrill** [comment]
@alanconway I defer to you to apply the `/lgtm` if you so wish to merge what is here

**Comment by alanconway** [comment]
/lgtm

I didn't intend to remove the TOC, I see one when I preview in firefox. Anyway this is all great stuff, so lets merge it and then tweak it for prettiness later.

**Comment by jcantrill** [comment]
@alanconway thoughts on if we need to start backporting these changes to 5.6?

**Pull request merged**
